File: carbon.ts
/**
 * Get information about carbon emission level for different energy types
 */
//% color=#004d00 weight=90 icon="\uf3fd" block="Carbon"
namespace carbon {

    /**
     * Sets carbon index value type
     * @param index target carbon index type; eg: CarbonIndex.moderate
     */
    //% help=carbon/set-carbon-index-value-type
    //% weight=100
    //% blockId=carbon_setCarbonIndexValueType
    //% block="carbon index level %index"
    //% shim=carbon::setCarbonIndexValueType
    export function setCarbonIndexValueType(index:CarbonIndex): number {
        return index;
    }

    /**
     * Get current carbon emission value
     */
    //% help=carbon/query-carbon-value
    //% weight=99
    //% blockId=carbon_queryCarbonValue
    //% block="get carbon value"
    //% shim=carbon::queryCarbonValue
    export function queryCarbonValue(): number {
        return Math.floor(Math.random() * 25) + 100;
    }

    /**
     * Get current carbon emission index level
     */
    //% help=carbon/query-carbon-index
    //% weight=98
    //% blockId=carbon_queryCarbonIndex
    //% block="get carbon index level"
    //% shim=carbon::queryCarbonIndex
    export function queryCarbonIndex(): number {
        return Math.floor(Math.random() * 25) + 100;
    }

    /**
     * Get current carbon emission index level as text
     */
    //% help=carbon/query-carbon-index-text
    //% weight=97
    //% blockId=carbon_queryCarbonIndexText
    //% block="get carbon index level as text"
    //% shim=carbon::queryCarbonIndexText
    export function queryCarbonIndexText(): string {
        let n = Math.floor(Math.random() * 5);
        
        switch(n)
        {
            case 0:
                return "very low";
            case 1:
                return "low";
            case 2:
                return "moderate";
            case 3:
                return "high";
            case 4:
                return "very high";
        }
        return "moderate";
    }

    /**
     * Get current carbon generation mix for specific type
     * @param mixType mixType; eg: GenerationMixType.gas
     */
    //% help=carbon/query-carbon-generation-mix
    //% weight=96
    //% blockId=carbon_queryCarbonGenerationMix
    //% block="get carbon generation mix|for type %mix"
    //% shim=carbon::queryCarbonGenerationMix
    export function queryCarbonGenerationMix(mixType : GenerationMixType): number
    {
        if(mixType == GenerationMixType.biomass){
            return Math.floor(Math.random() * 10) + 5;
        }
        if(mixType == GenerationMixType.coal){
            return Math.floor(Math.random() * 10) + 15;
        }
        if(mixType == GenerationMixType.imports){
            return Math.floor(Math.random() * 10) + 5;
        }
        if(mixType == GenerationMixType.gas){
            return Math.floor(Math.random() * 10) + 20;
        }
        if(mixType == GenerationMixType.nuclear){
            return Math.floor(Math.random() * 10) + 10;
        }
        if(mixType == GenerationMixType.hydro){
            return Math.floor(Math.random() * 10) + 1;
        }
        if(mixType == GenerationMixType.solar){
            return Math.floor(Math.random() * 10) + 10;
        }
        if(mixType == GenerationMixType.wind){
            return Math.floor(Math.random() * 10) + 2;
        }
        return 0;
    }
}

File: carbon.cpp
#include "pxt.h"
#include "CarbonService.h"

enum class GenerationMixType{
    gas = 0,
    coal = 1, 
    solar = 2, 
    wind = 3, 
    nuclear = 4, 
    hydro = 5, 
    biomass = 6, 
    imports = 7, 
    other = 8,
};

enum class CarbonIndex{
    very_low = 0,
    low = 1, 
    moderate = 2, 
    high = 3, 
    very_high = 4, 
};

static const char* const generationTypes[] = {"gas","coal","solar","wind","nuclear","hydro","biomass","imports","other"}; 
static const char* const carbonIndexValueTypes[] = {"very_low","low","moderate","high","very_high"}; 

/**
* share
* use
*/
PeridoRESTClient *radioTxRx = NULL;
CarbonService *carbonService = NULL;

//% color=#000000 weight=100 icon="\uf216" block="Carbon"
namespace carbon {

    int init() {
        if (radioTxRx == NULL)
            radioTxRx = new PeridoRESTClient(uBit.radio, uBit.messageBus, uBit.serial);

        if (carbonService == NULL)
            carbonService = new CarbonService(*radioTxRx);    

        return 0;
    }

    int stringToNumber(ManagedString input)
    {
        ManagedString resultString = "";
        int result = 0;

        int l = input.length();
        for(int x=0; x < l; ++x)
        {
            if(input.charAt(x) != '\n' && input.charAt(x) != '\r')
            {
                if(isdigit(input.charAt(x)) || input.charAt(x) == '.')
                {
                    resultString = resultString + input.charAt(x);
                }
                else
                {
                    result = -1;
                    break;
                }
            }
        }
        if(result > -1)
        {
            float f = atof(resultString.toCharArray());
            result = round(f);
        }
        return result;
    }

    //%
    int queryCarbonValue()
    {
        init();
        ManagedString reply = carbonService->getCarbonValue("value");
        int value = stringToNumber(reply);
        return value;
    }

    //Convert the carbon index from string to int
    // 'very low' 'low' 'moderate' 'high' 'very high'
    //  0          1     2          3      4
    //%
    int queryCarbonIndex()
    {
        init();

        ManagedString reply = carbonService->getCarbonIndex("index");

        int replyValue = -1;

        for(int loop=0; loop < 5; ++ loop)
        {
            //Compare the response text with the items in the array
            //if we get a match, the loop value is the correct number to return
            if(strcmp(carbonIndexValueTypes[loop], reply.toCharArray()) == 0)
            {
                replyValue = loop;
                break;
            }
        }
        return replyValue;
    }

    //%
    String queryCarbonIndexText()
    {
        init();

        ManagedString reply = carbonService->getCarbonIndex("index");
        return PSTR(reply);
    }

    //%
    int queryCarbonGenerationMix(GenerationMixType mix)
    {
        init();
        //Check the input value bounds
        if(mix < GenerationMixType::gas)
        {
            mix = GenerationMixType::gas;
        }

        if(mix > GenerationMixType::other)
        {
            mix = GenerationMixType::other;
        }
               
        ManagedString command = "genmix/";
        command = command + generationTypes[(int)mix];
        
        ManagedString reply = carbonService->getCarbonGenerationMix(command);
        int value = stringToNumber(reply);
        return value;
    }


    //%
    int setGenerationMixType(GenerationMixType GenerationMixType)
    {
        return (int)GenerationMixType;
    }

    //%
    int setCarbonIndexValueType(CarbonIndex index)
    {
        return (int)index;
    }

    
};
